Documentation/GPIO Pinning.md

```diff
@@ -2,174 +2,163 @@
+    - If the Group equals "E", then the GPIO Pin Number equals the Pad Number + 96
+    - If the Group equals "F", then the GPIO Pin Number equals the Pad Number + 120
+    - If the Group equals "G", then the GPIO Pin Number equals the Pad Number + 144
 
-##### Step 2b: Adding in the missing Resource Template
+- the format is "GPDX_IRQ", where X stands for a number. In that case the decimal GPIO Pin Number is X + 144 + 8
 
-If your device is unpinned, insert the following into the root of your device's entry:
+Now convert each of the GPIO Pin Numbers to their hexadecimal values (for instance using Calculator.app). These are your hexadecimal Candidate GPIO Pin numbers. Write these down.
 
-```
-    Name (SBFG, ResourceTemplate ()
-    {
-        GpioInt (Level, ActiveLow, ExclusiveAndWake, PullDefault, 0x0000,
-            "\\_SB.PCI0.GPI0", 0x00, ResourceConsumer, ,
-            )
-            {   // Pin list
-                0x0000
-            }
-    })
-```
+At the end of this step you have your ACPI ID and one or more hexadecimal Candidate GPIO numbers.
 
-You may now consider your device to be root pinned (but not well-root pinned) and can proceed to Step 2d.
+#### Step 2. Patching the DSDT
 
-##### Step 2c: Ensuring your device is well-pinned
+Before you patch your decompiled DSDT.dsl, make a copy of it so you can start over again when needed.
 
-Even if the number appearing in the above `Name` is `0x0`, it is still possible that your device is well-pinned. It is fairly easy to determine whether or not it is indeed well-pinned. Find the `_CRS` method of your device. If you can find a line that looks like this:
+##### Step 2.a The SBFG Name definition
 
+This section requires some familiarity with the structure and the syntax of dsl files. Syntax errors will result in a failure to compile the DSDT.
 
-```
-	Return (ConcatenateResTemplate (SBFB, SBFG))
-```
+Extract your DSTD to .dsl format and open it for manual editing. Find the section that contains the definitions for your touchpad by looking for "Device (YYYY)", where 'YYYY' must be replaced by the last four characters of your ACPI ID (for instance resulting in  "Device (ETPD)" or something similar).
+
+Check that this occurence is enclosed in a Scope definition which corresponds to the first parts of your ACPI ID to make sure that you have the right Device. Any changes that you will make using this guide are restricted to this Device section.
 
-Then your device is well-pinned. **Warning: this is not the same as the following:**
+Within this Device section, find a Name declaration for "SBFG", which should look like similar to:
 
 ```
-	Return (ConcatenateResTemplate (SBFB, SBFI))
+    Name (SBFG, ResourceTemplate ()
+    {
+	GpioInt (Level, ActiveLow, ExclusiveAndWake, PullDefault, 0x0000,
+	    "\\_SB.PCI0.GPI0", 0x00, ResourceConsumer, ,
+	    )
+	    {   // Pin list
+		0x0000
+	    }
+    })
 ```
 
-**you need `SBFG` and not `SBFI`**.
```

Comment:
/ Linux equivalent of FTE_i2c_write_cmd function
 IOReturn VoodooI2CFTETouchpadDriver::write_FTE_cmd(UInt16 reg, UInt16 cmd) {
     UInt16 buffer[] {
         reg,
         cmd
     };
     IOReturn retVal = kIOReturnSuccess;
     retVal = api->writeI2C(reinterpret_cast<UInt8*>(&buffer), sizeof(buffer));
     return retVal;
 }
 
 IOReturn VoodooI2CFTETouchpadDriver::message(UInt32 type, IOService* provider, void* argument)
 {
     switch (type)
     {
         case kKeyboardGetTouchStatus:
         {
 #if DEBUG
             IOLog("%s::getEnabledStatus = %s\n", getName(), ignoreall ? "false" : "true");
 #endif

Reply:
kext is feik